### PR TITLE
Try rendering again if a timed out tree receives an update

### PR DESCRIPTION
**Pull request by acdlite**
Found a bug related to suspending inside an already mounted tree. While investigating this I noticed we really don't have much coverage of suspended updates — nearly all are testing the initial mount. I think this would greatly benefit from some fuzz testing; still haven't thought of a good test case, though.

### CODE REVIEW

File: packages/jest-react/src/JestReact.js [hunk 1]
```diff
@@ -35,19 +35,17 @@ function assertYieldsWereCleared(root) {
 }
 
 export function toFlushAndYield(root, expectedYields) {
+  assertYieldsWereCleared(root);
```
Comment (acdlite) [inline packages/jest-react/src/JestReact.js]: I moved these outside because `captureAssertion` was swallowing the stack trace.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -1127,8 +1124,8 @@ function updateSuspenseComponent(
         // its fragment. We're going to skip over these entirely.
         const nextFallbackChildren = nextProps.fallback;
         const primaryChildFragment = createWorkInProgress(
-          currentFallbackChildFragment,
-          currentFallbackChildFragment.pendingProps,
+          currentPrimaryChildFragment,
+          currentPrimaryChildFragment.pendingProps,
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: This is a fix for an oopsies that could have been its own PR, but it was required to get the new test I wrote to pass. I can split it out if I need to.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 3]
```diff
@@ -1481,23 +1479,45 @@ function beginWork(
...
+            const primaryChildFragment: Fiber = (workInProgress.child: any);
+            const primaryChildExpirationTime =
+              primaryChildFragment.childExpirationTime;
+            if (
+              primaryChildExpirationTime !== NoWork &&
+              primaryChildExpirationTime <= renderExpirationTime
+            ) {
+              // The primary children have pending work. Use the normal path
+              // to attempt to render the primary children again.
+              return updateSuspenseComponent(
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: This is the main fix here.

File: packages/react-reconciler/src/ReactFiberSuspenseComponent.js [hunk 4]
```diff
@@ -23,3 +24,17 @@ export type SuspenseState = {|
   // `didTimeout` because it's not set unless the boundary actually commits.
   timedOutAt: ExpirationTime,
 |};
+
+export function shouldCaptureSuspense(
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberSuspenseComponent.js]: I split this out because it was getting hard to follow all the nested branches in the `throwException` function

File: packages/react-reconciler/src/ReactFiberUnwindWork.js [hunk 5]
```diff
@@ -407,9 +406,10 @@ function unwindWork(
         // Captured a suspense effect. Set the boundary's `alreadyCaptured`
         // state to true so we know to render the fallback.
         const current = workInProgress.alternate;
-        const currentState = current !== null ? current.memoizedState : null;
-        let nextState = workInProgress.memoizedState;
-        if (currentState === null) {
+        const currentState: SuspenseState | null =
+          current !== null ? current.memoizedState : null;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberUnwindWork.js]: I believe this will fix the issue in www, haven't confirmed yet. I'll try on my sandbox shortly.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 6]
```diff
@@ -1437,6 +1434,7 @@ function beginWork(
 ): Fiber | null {
   const updateExpirationTime = workInProgress.expirationTime;
 
+  // console.log(workInProgress.tag);
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: oops

**Comment by acdlite** [comment]
@sebmarkbage Hmm I remember hitting this issue with context and shrugging it off, not state though. (Oh, and context should just work now that the timed out children remain in the tree. Should add a test for that as a follow up.)
